05-mvc-with-rails/error-handling/readme.md

```diff
@@ -3,14 +3,388 @@
+- Compare and contrast the validation helpers `confirmation`, `inclusion`, `exclusion`, `length`, `presence`, `uniqueness`, and `numericality`
 - Compare and contrast the validation options `allow_nil`, `allow_blank`, and `on`
 - Create a custom validation on an ActiveRecord model
 
+# Validations
+
+## Framing
+
+In our Rails apps, we've been entering a lot of bogus data -- artists with blank names, for instance. That's fine while we're in development, but we don't want the users of our live app to be able to get away with that.
+
+#### How could we prevent users from entering blank data into a field?
+
+We could use Javascript, but that's easily circumvented with Chrome's Inspector.
+
+We could put that in the controller, like so:
+
+```rb
+if params[:name] == ""
+  redirect_to "/artists"
+end
+```
+
+...but that's going to lead to some pretty unwieldy controller methods. Besides, we may want there to be several routes that create or update an artist, and we'd need to copy and paste that bit of code for each one. That's hardly DRY!
+
+Wouldn't it be nice if, before an Artist record was saved, we could validate that the data entered into its fields were correct?
+
+## You do
+
+- Add the following code to your Artist model. What does each line do?
+
+```rb
+class Artist < ActiveRecord::Base
+  validates :name, uniqueness: true
+  validates :name, presence: true
+  validates :name, length: {in: 1..20}
+  validates :nationality, inclusion: ["USA", "Canada", "UK]
+end
+```
+
+- What's the *un*-DRY version of the following?
+
+```rb
+class Artist < ActiveRecord::Base
+  validates :name, :nationality, :photo_url, presence: true
+  validates :name, uniqueness: true, length: {maximum: 100}
+end
+```
+
+- Can you figure out what these would do?
+
+```rb
+class User < ActiveRecord::Base
+  validates :password, confirmation: true
+  validates :age, numericality: {only_integer: true, greater_than_or_equal_to: 13}
+  validates :country, exclusion: {in: ["North Korea"]}
+end
+```
+
+- What are the differences between the two following snippets?
+
+```rb
+class User < ActiveRecord::Base
+  validates :password, confirmation: true 
+  validates :ssn, uniqueness: true, allow_blank: true
+end
+
+class User < ActiveRecord::Base
+  validates :password, confirmation: true, on: create
+  validates :ssn, uniqueness: true, allow_nil: true
+end
+```
+
+## Custom validations
+
+You can also easily create your own custom validations. For instance, this will make Tunr reject any artist named Billy Ray Cyrus:
+
+```rb
+class Artist < ActiveRecord::Base
+  validate :break_billy_rays_achy_breaky_heart
+
+  def break_billy_rays_achy_breaky_heart
+    if self.name == "Billy Ray Cyrus"
+      errors.add(:name, "cannot be Billy Ray Cyrus, because he doesn't qualify as an artist.")
+    end
+  end
+end
+```
+
+#### Play around with this validation. What does `errors.add` do?
```

Comment:
`valid?` is what's happening under the hood - to show the effect of that - try creating a new repo on github without typing in a repo name. Full http request - with "name can't be blank"